Move CPROVER_{r,w,rw}_ok processing to back-end

tautschnig opened the pull request:
Please review commit-by-commit.

Fixes https://github.com/diffblue/cbmc/issues/7377.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

remi-delmas-3000 commented on src/util/pointer_expr.cpp:
```diff
@@ -227,13 +227,25 @@ void dereference_exprt::validate(
+exprt pointer_in_range_exprt::lower(const namespacet &ns) const
 {
   return and_exprt(
     {same_object(op0(), op1()),
      same_object(op1(), op2()),
```
Would these expressions be lowered in an ulterior lowering step ? (just thinking about what happen when passing these through `goto-instrument` multiple times).

EDIT: just saw that `same_object` is a function returning a more basic expression and not an `exprt`  subtype, you can disregard my comment.

remi-delmas-3000 commented on src/util/pointer_expr.cpp:
```diff
@@ -227,13 +227,25 @@ void dereference_exprt::validate(
+    {not_exprt{null_object(pointer())},
+     not_exprt{is_invalid_pointer_exprt{pointer()}},
+     not_exprt{deallocated(pointer(), ns)},
+     not_exprt{dead_object(pointer(), ns)},
```
This condition was not present in `good_pointer_def` which `r_or_rw_ok` is now supposed to replace (cf https://github.com/diffblue/cbmc/pull/7386/files)

tautschnig replied:
I believe this was a bug in `good_pointer_def`.

remi-delmas-3000 commented on src/util/pointer_expr.h:
```diff
@@ -386,7 +386,7 @@ class pointer_in_range_exprt : public ternary_exprt
 
   // translate into equivalent conjunction
-  exprt lower() const;
+  exprt lower(const namespacet &ns) const;
```
is this method inherited from some superclass  or trait, or is it added on a per-class basis ?

remi-delmas-3000 commented on src/solvers/flattening/bv_pointers.cpp:
```diff
@@ -203,6 +204,10 @@ literalt bv_pointerst::convert_rest(const exprt &expr)
     }
   }
+  else if(expr.id() == ID_r_ok || expr.id() == ID_w_ok || expr.id() == ID_rw_ok)
+    return convert(simplify_expr(to_r_or_w_ok_expr(expr).lower(ns), ns));
```
It seems we don't have any test that triggers this lowering step. Do we really get the guarantee that rw_ok expressions are going to be converted only when going to the solver ? (assuming this is what this PR tries to achieve ?)

tautschnig replied:
Should now be done, the early-expansion removal wasn't there back then.

remi-delmas-3000 commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2499,6 +2499,14 @@ void smt2_convt::convert_expr(const exprt &expr)
   }
+  else if(expr.id() == ID_r_ok || expr.id() == ID_w_ok || expr.id() == ID_rw_ok)
+  {
+    convert_expr(simplify_expr(to_r_or_w_ok_expr(expr).lower(ns), ns));
```
no coverage here,  could this be a sign that these expressions still get lowered before reaching that step ?

tautschnig replied:
This should now be ok - at the time of you viewing this I had not included the removal of early expansion.

jimgrundy commented:
Looks like we need to add __CPROVER_is_in_range to the mix

tautschnig committed:
Remove {r,w,rw}_ok and pointer_in_range expansion from C front-end

These expressions are now rewritten to their prophecy_* counterparts by
goto-program conversion, and back-ends directly handle these prophecy_*
expressions. Unconditional rewrites by goto_check_ct (even when no
checks are enabled) are no longer necessary.

Fixes: #7377

tautschnig committed:
r/w/rw_ok and pointer_in_range depend on deallocated and dead_object

The evaluation of these expressions depends on the specific values of
__CPROVER_deallocated and __CPROVER_dead_object at the time the
expression is used. Therefore, create new prophecy_* variants of these
expressions to retain the properties expected of an expression (as
opposed to a statement/function call).

These prophecy_* expressions can safely be passed on to back-ends, be
simplified, and do not need to be rewritten during instrumentation. For
now, they are just handled by lowering the expressions (in the SAT and
SMT back-ends). We may, in future, decide to have back-end specific (and
possibly more efficient) handling of these.

tautschnig commented:
> Looks like we need to add __CPROVER_is_in_range to the mix

This is done (including a few fixes specific to just that).

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2684,8 +2683,15 @@ exprt c_typecheck_baset::do_special_functions(
 
+    symbol_exprt deallocated =
+      lookup(CPROVER_PREFIX "deallocated").symbol_expr();
+    symbol_exprt dead = lookup(CPROVER_PREFIX "dead_object").symbol_expr();
```
There's something wrong here. The connection to `deallocated` is very much a mechanism for implementing the predicate, and that shouldn't happen in the front-end.

tautschnig replied:
Well, I really just moved (while also making it more obvious) something from `goto_check_ct` to type checking. Both of which are in the front-end...

peterschrammel replied:
Would it help if we lowered the side-effecty function to a predicate in the built-in library?
The C implementation of the function can refer to the global state to call the built-in predicate.

kroening reviewed (changes requested):
There's a fundamental architectural problem with this PR.
Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

tautschnig commented:
> There's a fundamental architectural problem with this PR. Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

I agree that there is an architectural problem, but I'm not convinced it's in this PR: this PR just makes it very obvious: as they are (before this PR), `{r,w,rw}_ok_exprt` and `pointer_in_range_exprt` are not actually expressions: two occurrences with the same set of (previous) arguments may still result in different evaluations. Such "expressions" we would otherwise prefix with `side_effect_`.

jimgrundy commented:
I'm in agreement with Michael. The prior API aligns to the behavior one wishes these primitives had, but not the behavior they actually have. This dependence on the DEAD and DEALLOCATED values is not desirable, but also very real, and the way it is obscured by the present API has proved confusing. I think it best that the API make the behavior plain.

That's not to say that we wouldn't one day aspire to improving the behavior, perhaps by using SMT more in the encoding, and at that time setting the API back to what it was before. I'd very much like to see that.

thomasspriggs commented:
I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -

1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.

tautschnig commented:
> I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> 
>     1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> 
>     2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.

Thank you for the guidance, I'll implement this.

tautschnig commented:
> There's a fundamental architectural problem with this PR. Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

@kroening Would you mind taking another look at the implementation, specifically in the commit [r/w/rw_ok and pointer_in_range depend on deallocated and dead_object](https://github.com/diffblue/cbmc/pull/7395/commits/1ba2ee65c85958d414109e988f82e3409db795fc)?

kroening commented on src/goto-programs/builtin_functions.cpp:
```diff
@@ -696,7 +696,10 @@ void goto_convertt::do_havoc_slice(
+  symbol_exprt deallocated =
+    ns.lookup(CPROVER_PREFIX "deallocated").symbol_expr();
+  symbol_exprt dead = ns.lookup(CPROVER_PREFIX "dead_object").symbol_expr();
+  w_ok_exprt ok_expr{arguments[0], arguments[1], deallocated, dead};
```
Do you still need this change?

tautschnig replied:
It won’t be subject to cleaning at this point, but I could of course explicitly invoke cleaning.

kroening commented:
I have on remaining problem with this PR.

There are now two variants of the memory predicates: prior to filling in the particular realisation using `dead` and `deallocated`, and post. They are very different, and in particular, there may well be many different realisations.

May I thus advise to make this separate expression classes, with different arities, and different identifiers -- much like the trick used in the CHC encoder. Right now this is very difficult to understand by looking at the code, or the classes.

tautschnig commented:
> May I thus advise to make this separate expression classes, with different arities, and different identifiers -- much like the trick used in the CHC encoder. Right now this is very difficult to understand by looking at the code, or the classes.

Any preference on renaming the front end vs the to-be-back-end ones? We could have ones with a “frontend” in their name like we do for some codet classes.

kroening commented:
> Any preference on renaming the front end vs the to-be-back-end ones? We could have ones with a “frontend” in their name like we do for some codet classes.

We have three categories of expressions:
1. Ones that only exist in some of the language front-ends, but never make their way into the GOTO binary. We have used the`frontend_` prefix for those.
2. Ones that exist in the GOTO binary, and may in fact have an implicit "program state" argument.
3. Proper logic ones. No program state.

The one without the extra arguments fall into category 2. Once you add the arguments you have an expression in category 3. I'd use `r/w_ok` (no prefix) for the ones prior to baking in the state. We have sometimes used the prefix `mathematical_` to identify expressions and types of the third kind, and to distinguish these from their program counterparts.

The variant with the extra arguments is a specific realisation of the predicates that uses prophecy variables. How about prefixing them with `prophecy_`.

martin-cs commented:
@tautschnig I still have a "review requested" for this.  Given the discussion and the conflicting files, I will rely on you saying when you want it reviewed.

tautschnig commented:
> > I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> > ```
> > 1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> > 
> > 2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.
> > ```
> 
> Thank you for the guidance, I'll implement this.

This is now done.

tautschnig commented:
> The variant with the extra arguments is a specific realisation of the predicates that uses prophecy variables. How about prefixing them with `prophecy_`.

I believe this is now implemented as suggested. Could I please get a re-review?

thomasspriggs commented:
> > > I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> > > ```
> > > 1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> > > 
> > > 2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.
> > > ```
> > 
> > 
> > Thank you for the guidance, I'll implement this.
> 
> This is now done.

Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039

The changes you made to the `smt2_incremental_decision_procedure` look fine to me. Just FYI myself and @esteffin are each working on changes which are likely to touch the same lines of code where the decision procedure does the lowering. So there may be a need for some minor rebasing to resolve merge conflicts between the 3 of us in the near future.

thomasspriggs commented on src/ansi-c/expr2c_class.h:
```diff
@@ -28,6 +28,7 @@ class qualifierst;
 class namespacet;
 class r_or_w_ok_exprt;
 class pointer_in_range_exprt;
+class r_or_w_ok_exprt;
```
:pick: This change doesn't seem to match the commit message, as the commit message states it is adding a method, but the change is forward declaring a class.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -1464,6 +1464,24 @@ static smt_termt convert_expr_to_smt(
+  const prophecy_r_or_w_ok_exprt &prophecy_r_or_w_ok,
+  const sub_expression_mapt &converted)
+{
+  UNIMPLEMENTED_FEATURE(
```
Nitpick: With the lowing added to the decision procedure, it is not so much that this is an unimplemented feature but that that should be `UNREACHABLE_BECAUSE("prophecy_r_or_w_ok expression should have been lowered by the decision procedure before conversion to smt terms.")`

tautschnig replied:
Thank you, has been fixed.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -1464,6 +1464,24 @@ static smt_termt convert_expr_to_smt(
+  const prophecy_pointer_in_range_exprt &prophecy_pointer_in_range,
+  const sub_expression_mapt &converted)
+{
+  UNIMPLEMENTED_FEATURE(
```
See previous `UNIMPLEMENTED_FEATURE` nitpick.

tautschnig replied:
Fixed according to your suggestion.

tautschnig commented:
> Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - [thomasspriggs@9780929](https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039)

These were failing in two of the regression tests only. I'll happily try this again, but there may be a need for more debugging.

thomasspriggs commented:
> > Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - [thomasspriggs@9780929](https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039)
> 
> These were failing in two of the regression tests only. I'll happily try this again, but there may be a need for more debugging.

:thinking: All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - https://github.com/diffblue/cbmc/pull/7730

tautschnig commented:
> 🤔 All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730

Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.

thomasspriggs commented:
> > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> 
> Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.

I see. I hadn't realised that this was only failing on some platforms.

thomasspriggs commented:
> > > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> > 
> > 
> > Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.
> 
> I see. I hadn't realised that this was only failing on some platforms.

Ok, it looks like removing the `simplify_expr` step fixes the 2 failing tests of the mac job - https://github.com/diffblue/cbmc/pull/7730/commits/79f95f8e43e7341dc1a11d04f889f03a22cc3c18 I have a hunch that the simplification is simplifying away expressions which need to be defined within the decision procedure for the `get` function to work as it should do, but I don't have a mac on hand to confirm right now.

That means the issue with `dynamic_sizeof1` and the coverage job is still a mystery for the moment.

tautschnig commented:
> > > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> > 
> > 
> > Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.
> 
> I see. I hadn't realised that this was only failing on some platforms.

I believe I have root-caused this to https://github.com/diffblue/cbmc/pull/7735 (~~which is now also included in this PR~~ which is now merged).

kroening commented:
This needs another nudge. At the moment, the lowering of X_ok to prophecy_X_ok is done during goto-conversion, which precludes any other approach. This should be done by a separate pass, appropriate for the analyser.

kroening commented:
The contract instrumenter in goto-instrument should then generate X_ok, not prophecy_X_ok.

tautschnig committed:
Move rw_ok rewriting to separate pass

This reverts selected changes from "r/w/rw_ok and pointer_in_range
depend on deallocated and dead_object" for transformations of goto
programs can now safely introduce rw_ok expressions into the goto
program. Not all analyses, however, are able to cope with these
expressions. Dependence graphs (and any analyses building upon these),
for example, also require this rewriting. goto-instrument is adjusted
accordingly.

tautschnig commented:
> This needs another nudge. At the moment, the lowering of X_ok to prophecy_X_ok is done during goto-conversion, which precludes any other approach. This should be done by a separate pass, appropriate for the analyser.

I have now moved the lowering to a separate pass called `rewrite_rw_ok`.

> The contract instrumenter in goto-instrument should then generate X_ok, not prophecy_X_ok.

Done. I have added that new pass as an additional commit, which partly reverts some of the earlier changes. So it might make sense to review the overall diff rather than individual commits on this occasion.

remi-delmas-3000 commented on regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc:
is this meant to be in this commit ?

remi-delmas-3000 commented on regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc:
is this meant to be in this commit ?

remi-delmas-3000 commented on regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc:
is this meant to be here ?

tautschnig replied:
Yes: the intermediate version that did the expansion during goto-program conversion resulted in goto programs that have non-assert additional instructions (because of side-effect cleanup). This, in turn, affects what asserts get labelled as preconditions. So, really, it just changed the labels on some properties. Moving the expansion into a separate rewrite step undid this change.

remi-delmas-3000 commented on regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc:
is this meant to be here ?

tautschnig replied:
Same as above: just a change in property labelling that was no longer in place after moving to a separate rewrite step.

remi-delmas-3000 commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -406,6 +406,7 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
   PARSE_OPTIONS_GOTO_TRACE(cmdline, options);
 
   // Options for process_goto_program
+  options.set_option("rewrite-rw-ok", true);
```
why is this new encoding switchable ? and what happens when it is switched off ?

tautschnig replied:
When this is switched off, the goto program will contain `{r,w,rw}_ok_exprt`, and the analysis needs to be able to deal with these. The CHC-based approach, for example, does have its own way of handling them.

remi-delmas-3000 commented on src/goto-programs/rewrite_rw_ok.h:
Could you add some documentation of when this rewriting must take place or when it may be omitted ? I still have a hard time figuring out the lifecycle of these expressions. The ansi-c front end convers rw_ok expressions to the prophecy versions right away, and then these get eliminated right before symex with CBMC, but may or may not remain in the model when other tools are invoked ?

tautschnig replied:
I have added another sentence to this file's documentation.

NlightNFotis commented on src/ansi-c/expr2c_class.h:
```diff
@@ -28,6 +28,7 @@ class qualifierst;
 class namespacet;
 class r_or_w_ok_exprt;
 class pointer_in_range_exprt;
+class r_or_w_ok_exprt;
```
This is present in line 29 as well. Safe to assume this was a squashing/rebasing mistake? Or is there a particular reason for why this was added here?

NlightNFotis replied:
There's also the question of the commit message mentioning:

> expr2c/{r,w,rw}_ok: add a dedicated method

But I don't see any method being added here. Is the commit message in sync with the intended change here, or do I misunderstand the intent?

NlightNFotis replied:
**EDIT**: Yeah, when I viewed all of the files in the unified file changes tab, it seems that the change isn't there, meaning that the addition was removed at another commit and I missed it.

That probably means that the whole of the first commit is redundant, if I'm not mistaken.

tautschnig replied:
Thank you for this careful review! Indeed, the bulk of the first commit was already merged in #7460, so I just need to drop this commit.

NlightNFotis commented on src/util/pointer_expr.cpp:
```diff
@@ -238,3 +238,30 @@ exprt pointer_in_range_exprt::lower() const
+
+exprt prophecy_pointer_in_range_exprt::lower(const namespacet &ns) const
+{
+  return and_exprt(
```
As far as I can tell, both `conjunction` and the `and_exprt` constructor here should behave in a functionally equivalent manner.

If my understanding is correct, is there any particular reason for why the two similar methods use a different method for the same functionality? 

If not, does it make sense to refactor them to align their internals in a way that shows the functions' equivalence (from a mechanical standpoint, I can see they're adding different conjuncts to the end expression).

tautschnig replied:
Oh, I have no idea why I wasn't consistent here -- these are indeed equivalent, except `conjunction` will have an extra layer of indirection. Now consistently using `and_exprt` here.

NlightNFotis commented on src/goto-programs/goto_clean_expr.cpp:
```diff
@@ -73,6 +74,15 @@ bool goto_convertt::needs_cleaning(const exprt &expr)
   }
 
+  // Rewrite rw_ok and pointer_in_range front-end expressions into ones
+  // including prophecy expressions.
```
The comment here would be better located around line 449, if I am not mistaken?

Reading it here made me assume that this function would do the cleanup, whereas in reality it just marks expressions for cleanup. The actual rewriting happens later, around line 449 as far as I can tell, correct?

NlightNFotis replied:
**EDIT**: Ignore my comment here - I saw it got removed in a later commit when this check was removed.

NlightNFotis commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1767,6 +1768,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   {
     do_indirect_call_and_rtti_removal();
     do_remove_returns();
+    rewrite_rw_ok(goto_model);
```
I see only the `full-slice` option is rewriting the `rw` expressions.

Is there some specific reason the other slices don't want to display the same functionality? Apologies, but I couldn't tell the intent here just by reading the code.

tautschnig replied:
Other forms of slicing don't use expression-level granularity. I have updated the commit message to include this information.

NlightNFotis reviewed (commented):
Some questions before I hit approve.

feliperodri commented:
@NlightNFotis could you give a final look?

TGWDB commented:
> @NlightNFotis could you give a final look?

If you specifically want Fotis it will take a couple of weeks, but others here can take a look much sooner than that.

tautschnig commented:
> > @NlightNFotis could you give a final look?
> 
> If you specifically want Fotis it will take a couple of weeks, but others here can take a look much sooner than that.

I believe I have addressed all of Fotis's comments, so it should likely be just any code owner for the incremental SMT2 back-end. Will update assignments accordingly.

thomasspriggs reviewed (approved):
Approving for the incremental SMT2 changes.

martin-cs reviewed (approved):
I believe that the changes have been implemented correctly and do not raise any issues in the parts of the code that I am "owner" for.

I am as yet unsure why this is needed or why this is a good design to meet that need.  However, judging by the comment history, this has been considered and it is something that I should resolve with a chat with @tautschnig rather than via github.

tautschnig commented:
> I am as yet unsure why this is needed or why this is a good design to meet that need. However, judging by the comment history, this has been considered and it is something that I should resolve with a chat with @tautschnig rather than via github.

Let me try to clarify publicly nonetheless (the why-is-this-needed part):
1) Not all middle-ends agree on how `__CPROVER_{r,w,rw}_ok` ought to be processed. Notably, the `cprover` tool has its own way of handling these. So we shouldn't prematurely expand them in the C front-end.
2) Early expansion leads to redundant checks, which result in a performance penalty. See #7378 (and related issues linked from there).
3) In the status before the changes in this PR, this early expansion was necessary (as not all middle-ends were able to cope with these expressions), and this was silently done as part of the property instrumentation code (even when no property checks were requested). This is surprising behaviour.
